waspc/data/Generator/templates/react-app/src/actions/index.js

```diff
@@ -0,0 +1,152 @@
+import {
+  useMutation,
+  useQueryClient,
+} from 'react-query'
+
+export { configureQueryClient } from '../queryClient'
+
+/**
+ * A hook for adding extra behavior to a Wasp Action (e.g., optimistic updates).
+ *
+ * @param actionFn The Wasp Action you wish to enhance/decorate.
+ * @param {Object} actionOptions An options object for enhancing/decorating the given Action.
+ * @returns {Function} A decorated Action with added behavior but an unchanged API.
+ */
+export function useAction(actionFn, actionOptions) {
+  const queryClient = useQueryClient();
+
+  let options = {}
+  if (actionOptions?.optimisticUpdates) {
+    const optimisticUpdateDefinitions = actionOptions.optimisticUpdates.map(translateToInternalDefinition)
+    options = makeRqOptimisticUpdateOptions(queryClient, optimisticUpdateDefinitions)
+  }
+
+  // NOTE: We decided to hide React Query's extra mutation features (e.g.,
+  // isLoading, onSuccess and onError callbacks, synchronous mutate) and only
+  // expose a simple async function whose API matches the original Action.
+  const mutation = useMutation(actionFn, options)
+  return (args) => mutation.mutateAsync(args)
+}
+
+/**
+ * Translates/Desugars a public optimistic update definition object into a definition object our
+ * system uses internally.
+ * 
+ * @param {Object} publicOptimisticUpdateDefinition An optimistic update definition object that's a part of the public API:
+ * https://wasp-lang.dev/docs/language/features#the-useaction-hook.
+ * @returns {Object} An internally-used optimistic update definition object.
+ */
+function translateToInternalDefinition(publicOptimisticUpdateDefinition) {
+  const { getQuerySpecifier, updateQuery } = publicOptimisticUpdateDefinition
+
+  const definitionErrors = []
+  if (typeof getQuerySpecifier !== 'function') {
+    definitionErrors.push('`getQuerySpecifier` is not a function.')
+  }
+  if (typeof updateQuery !== 'function') {
+    definitionErrors.push('`updateQuery` is not a function.')
+  }
+  if (definitionErrors.length) {
+    throw new TypeError(`Invalid optimistic update definition: ${definitionErrors.join(', ')}.`)
+  }
```

Comment:
I'm still not sure how to handle these things in our API (by these things I mean runtime type checks for the stuff users pass into our functions). Should we just pretend we're getting the right values all the time and let the system break by itself? Should we fail on the first error we find, or collect as much as possible before throwing an exception.

Seems like our codebase takes several different approaches to this issue.

Reply:
Yeah, good question. I think what you are doing, validating our expectations and then showing appropriate errors, makes sense here. I think I did something similar for the Google config function return object shape.

If it is something like a user-defined function that just explodes, and we have no recourse, I have just usually let them explode so they get contextual stack traces. But thinking more about it holistically does make sense.

Our future change to TS should help with these types of checks though, right?